x-pack/plugins/task_manager/server/lib/retryable_bulk_update.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SavedObjectError } from '@kbn/core-saved-objects-common';
+import { ConcreteTaskInstance } from '../task';
+import { TaskStore, BulkUpdateResult, BulkGetResult } from '../task_store';
+import { isErr, isOk, asErr } from './result_type';
+import { BulkUpdateTaskResult } from '../task_scheduling';
+
+export const MAX_RETRIES = 2;
+
+export interface RetryableBulkUpdateOpts {
+  taskIds: string[];
+  getTasks: (taskIds: string[]) => Promise<BulkGetResult>;
+  filter: (task: ConcreteTaskInstance) => boolean;
+  map: (task: ConcreteTaskInstance) => ConcreteTaskInstance;
+  store: TaskStore;
+}
+
+export async function retryableBulkUpdate({
+  taskIds,
+  getTasks,
+  filter,
+  map,
+  store,
+}: RetryableBulkUpdateOpts): Promise<BulkUpdateTaskResult> {
+  const resultMap: Record<string, BulkUpdateResult> = {};
+  const tasksToUpdate = (await getTasks(taskIds))
+    .reduce<ConcreteTaskInstance[]>((acc, task) => {
+      if (isErr(task)) {
+        resultMap[task.error.id] = buildBulkUpdateErr(task.error);
+      } else {
+        acc.push(task.value);
+      }
+      return acc;
+    }, [])
+    .filter(filter)
+    .map(map);
+  const bulkUpdateResult = await store.bulkUpdate(tasksToUpdate);
+  for (const result of bulkUpdateResult) {
+    const taskId = getId(result);
+    resultMap[taskId] = result;
+  }
+
+  let retry = 1;
+  while (retry++ <= MAX_RETRIES && getRetriableTaskIds(resultMap).length > 0) {
+    const retriesToGet = getRetriableTaskIds(resultMap);
+    const retriesToUpdate = (await getTasks(retriesToGet))
```

Comment:
nit: this code block and the above code block look very similar. could they be combined into a helper function?

Reply:
@ymao1 good point! I was able to remove repetition by creating a closure. See https://github.com/elastic/kibana/pull/148985/commits/9c096e5662e8bede083dcbe4caf5b24f9ca3ee39?diff=unified&w=1. Could you take a look to make sure it looks good?

Reply:
LGTM!